From 59e9df8eb0c28f5c9e905f753ee3166111e5ec19 Mon Sep 17 00:00:00 2001
From: Liam Howlett <Liam.Howlett@oracle.com>
Date: Thu, 24 Nov 2022 10:50:21 -0500
Subject: [PATCH 18/58] maple_tree: Protect mas_prev_nentry() from dead nodes

Don't use the pivot array before checking if the node is dead.  This is
for safe use in rcu mode.  This is necessary for the RCU mode of the
maple tree.

Fixes: 54a611b60590 ("Maple Tree: add new data structure")
Signed-off-by: Liam Howlett <Liam.Howlett@oracle.com>
Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 lib/maple_tree.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/maple_tree.c b/lib/maple_tree.c
index 502d83d57..ca89869eb 100644
--- a/lib/maple_tree.c
+++ b/lib/maple_tree.c
@@ -4856,6 +4856,11 @@ static inline void *mas_prev_nentry(struct ma_state *mas, unsigned long limit,
 		goto retry;
 	}
 
+	if (offset == mt_pivots[mt])
+		pivot = mas->max;
+	else
+		pivot = pivots[offset];
+
 	while (offset && ((!mas_slot(mas, slots, offset) && pivot >= limit) ||
 	       !pivot))
 		pivot = pivots[--offset];
-- 
2.39.0.rc2.1.gbd5df96b79

